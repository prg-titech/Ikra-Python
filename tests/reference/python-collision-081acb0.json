{
    "basename": "python-collision-081acb0",
    "cmd": "new-src/interface.py --emit-py {infile}",
    "infile": "examples/collision/collision.py",
    "infile_hash": "32b6014a5cbf9597cb4b52c12e91d20c646ad1a3138607b7120da6e7",
    "stdout": "python-collision-081acb0.stdout",
    "stdout_hash": "35dc2947d9c5e7944e379ecbe01f8f0acd6452f7aef7802700a8f60c",
    "stderr": null,
    "stderr_hash": null,
    "returncode": 0
}