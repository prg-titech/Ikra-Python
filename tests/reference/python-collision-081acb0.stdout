pygame 2.0.1 (SDL 2.0.14, Python 3.7.5)
Hello from the pygame community. https://www.pygame.org/contribute.html

from __future__ import annotations
import os, time, math, pygame, random
from sanajeh import DeviceAllocator
kSeed: int = 42
kMaxMass: float = 500
kDt: float = 0.01
kGravityConstant: float = 4e-06
kDampeningFactor: float = 0.05
kMergeThreshold: float = 0.1
kTimeInterval: float = 0.05

class Vector():

    def __init__(self, x_: float, y_: float):
        self.x: float = x_
        self.y: float = y_

    def add(self, other: Vector) -> Vector:
        self.x += other.x
        self.y += other.y
        return self

    def plus(self, other: Vector) -> Vector:
        return Vector((self.x + other.x), (self.y + other.y))

    def subtract(self, other: Vector) -> Vector:
        self.x -= other.x
        self.y -= other.y
        return self

    def minus(self, other: Vector) -> Vector:
        return Vector((self.x - other.x), (self.y - other.y))

    def scale(self, ratio: float) -> Vector:
        self.x *= ratio
        self.y *= ratio
        return self

    def multiply(self, multiplier: float) -> Vector:
        return Vector((self.x * multiplier), (self.y * multiplier))

    def divide_by(self, divisor: float) -> Vector:
        self.x /= divisor
        self.y /= divisor
        return self

    def divide(self, divisor: float) -> Vector:
        return Vector((self.x / divisor), (self.y / divisor))

    def dist_origin(self) -> float:
        return math.sqrt(((self.x * self.x) + (self.y * self.y)))

    def to_zero(self) -> Vector:
        self.x = 0.0
        self.y = 0.0
        return self

class Body():

    def __init__(self):
        self.random_state_: DeviceAllocator.RandomState = None
        self.merge_target_ref: Body = None
        self.pos_x: float = 0
        self.pos_y: float = 0
        self.vel_x: float = 0
        self.vel_y: float = 0
        self.force_x: float = 0
        self.force_y: float = 0
        self.mass: float = 0
        self.has_incoming_merge: bool = False
        self.successful_merge: bool = False

    def Body(self, idx: int):
        random.seed(idx)
        self.merge_target_ref = None
        self.pos_x = ((2.0 * random.uniform(0, 1)) - 1.0)
        self.pos_y = ((2.0 * random.uniform(0, 1)) - 1.0)
        self.vel_x = ((random.uniform(0, 1) - 0.5) / 1000)
        self.vel_y = ((random.uniform(0, 1) - 0.5) / 1000)
        self.force_x = 0.0
        self.force_y = 0.0
        self.mass = (((random.uniform(0, 1) / 2) + 0.5) * kMaxMass)
        self.has_incoming_merge = False
        self.successful_merge = False

    def compute_force(self):
        self.force_x = 0.0
        self.force_y = 0.0
        DeviceAllocator.device_do(Body, Body.apply_force, self)

    def apply_force(self, other: Body):
        if (other is not self):
            d_x: float = (self.pos_x - other.pos_x)
            d_y: float = (self.pos_y - other.pos_y)
            dist: float = math.sqrt(((d_x * d_x) + (d_y * d_y)))
            f: float = (((kGravityConstant * self.mass) * other.mass) / ((dist * dist) + kDampeningFactor))
            __auto_v0_x: float = (d_x * f)
            __auto_v0_y: float = (d_y * f)
            __auto_v1_x: float = (__auto_v0_x / dist)
            __auto_v1_y: float = (__auto_v0_y / dist)
            other.force_x += __auto_v1_x
            other.force_y += __auto_v1_y

    def body_update(self):
        __auto_v0_x: float = (self.force_x * kTimeInterval)
        __auto_v0_y: float = (self.force_y * kTimeInterval)
        __auto_v1_x: float = (__auto_v0_x / self.mass)
        __auto_v1_y: float = (__auto_v0_y / self.mass)
        self.vel_x += __auto_v1_x
        self.vel_y += __auto_v1_y
        __auto_v2_x: float = (self.vel_x * kTimeInterval)
        __auto_v2_y: float = (self.vel_y * kTimeInterval)
        self.pos_x += __auto_v2_x
        self.pos_y += __auto_v2_y
        if ((self.pos_x < (- 1)) or (self.pos_x > 1)):
            self.vel_x = (- self.vel_x)
        if ((self.pos_y < (- 1)) or (self.pos_y > 1)):
            self.vel_y = (- self.vel_y)

    def check_merge_into_this(self, other: Body):
        if ((not other.has_incoming_merge) and (self.mass > other.mass)):
            d_x: float = (self.pos_x - other.pos_x)
            d_y: float = (self.pos_y - other.pos_y)
            dist_square: float = math.sqrt(((d_x * d_x) + (d_y * d_y)))
            if (dist_square < (kMergeThreshold * kMergeThreshold)):
                self.merge_target_ref = other
                other.has_incoming_merge = True

    def initialize_merge(self):
        self.merge_target_ref = None
        self.has_incoming_merge = False
        self.successful_merge = False

    def prepare_merge(self):
        DeviceAllocator.device_do(Body, Body.check_merge_into_this, self)

    def update_merge(self):
        m: Body = self.merge_target_ref
        if (m is not None):
            if (m.merge_target_ref is not None):
                new_mass: float = (self.mass + m.mass)
                __auto_v0_x: float = (m.vel_x * m.mass)
                __auto_v0_y: float = (m.vel_y * m.mass)
                __auto_v1_x: float = (self.vel_x * self.mass)
                __auto_v1_y: float = (self.vel_y * self.mass)
                __auto_v2_x: float = (__auto_v1_x + __auto_v0_x)
                __auto_v2_y: float = (__auto_v1_y + __auto_v0_y)
                new_vel_x: float = (__auto_v2_x / new_mass)
                new_vel_y: float = (__auto_v2_y / new_mass)
                m.mass = new_mass
                m.vel_x = new_vel_x
                m.vel_y = new_vel_y
                self.pos_x += m.pos_x
                self.pos_y += m.pos_y
                __auto_v3_x: float = self.pos_x
                __auto_v3_y: float = self.pos_y
                m.pos_x = (__auto_v3_x / 2)
                m.pos_y = (__auto_v3_y / 2)
                self.successful_merge = True

    def delete_merged(self):
        if self.successful_merge:
            DeviceAllocator.destroy(self)

def kernel_initialize_bodies():
    DeviceAllocator.device_class(Body)

def _update():
    DeviceAllocator.parallel_do(Body, Body.compute_force)
    DeviceAllocator.parallel_do(Body, Body.body_update)
    DeviceAllocator.parallel_do(Body, Body.initialize_merge)
    DeviceAllocator.parallel_do(Body, Body.prepare_merge)
    DeviceAllocator.parallel_do(Body, Body.update_merge)
    DeviceAllocator.parallel_do(Body, Body.delete_merged)

def main(allocator, do_render):
    '\n  Rendering settings\n  '

    def render(b):
        size = pow((b.mass / 5), 0.125)
        px = int(((((b.pos.x / 2) + 0.5) * 500) - (size / 2)))
        py = int(((((b.pos.y / 2) + 0.5) * 500) - (size / 2)))
        pygame.draw.circle(screen, (255, 255, 255), (px, py), size)
    if do_render:
        os.environ['SDL_VIDEODRIVER'] = 'x11'
        screen_width = 500
        screen_height = 500
        pygame.init()
        screen = pygame.display.set_mode((screen_width, screen_height))
        pygame.display.flip()
    num: int = 100
    iter: int = 5000
    allocator.initialize()
    initialize_time = time.perf_counter()
    allocator.parallel_new(Body, num)
    parallel_new_time = time.perf_counter()
    for x in range(iter):
        allocator.parallel_do(Body, Body.compute_force)
        allocator.parallel_do(Body, Body.body_update)
        allocator.parallel_do(Body, Body.initialize_merge)
        allocator.parallel_do(Body, Body.prepare_merge)
        allocator.parallel_do(Body, Body.update_merge)
        allocator.parallel_do(Body, Body.delete_merged)
        if do_render:
            allocator.do_all(Body, render)
            pygame.display.flip()
            screen.fill((0, 0, 0))
    end_time = time.perf_counter()
    print(('parallel new time(%-5d objects): %.dµs' % (num, ((parallel_new_time - initialize_time) * 1000000))))
    print(('average computation time: %dµs' % (((end_time - parallel_new_time) * 1000000) / iter)))
    print(('overall computation time(%-4d iterations): %dµs' % (iter, ((end_time - parallel_new_time) * 1000000))))

